File: cpp/src/pwd.cpp
#include "types.hpp"

#include "util.hpp"

namespace filefs::detail {

bool FileSystemState::init_pwdtmp(const std::string_view s) {
  pwd_tmp.assign(s.data(), s.size());
  return true;
}

bool FileSystemState::add_to_pwdtmp(const std::size_t /*pathsize*/, const std::string_view s) {
  if (s == ".") {
    return true;
  }

  if (s == "..") {
    for (std::size_t i = 1; i < pwd_tmp.size(); ++i) {
      if (pwd_tmp[pwd_tmp.size() - i - 1] == '/') {
        pwd_tmp.resize(pwd_tmp.size() - i);
        return true;
      }
    }
    return false;
  }

  pwd_tmp.append(s.data(), s.size());
  pwd_tmp.push_back('/');
  return true;
}

}  // namespace filefs::detail

namespace filefs {

bool FileSystem::chdir(const std::string_view pathname) {
  if (!state_ || !state_->mounted()) {
    return false;
  }

  const int len_n = static_cast<int>(pathname.size());
  std::uint32_t blockindex = 0U;
  int start = 0;
  if (!pathname.empty() && pathname.front() == '/') {
    blockindex = 1U;
    start = 1;
    if (!state_->init_pwdtmp("/")) {
      return false;
    }
  } else {
    if (state_->tmp.state == 0U) {
      blockindex = state_->pwd_blockindex;
      if (!state_->init_pwdtmp(state_->pwd)) {
        return false;
      }
    } else {
      blockindex = state_->tmp.pwd_blockindex;
      if (!state_->init_pwdtmp(state_->tmp.pwd)) {
        return false;
      }
    }
  }

  std::array<char, BlockNameMaxSize + 2> s{};
  int slen = 0;
  for (int i = start; i < len_n; ++i) {
    if (pathname[static_cast<std::size_t>(i)] == '/') {
      if (slen == 0) {
        continue;
      }
      const std::string_view comp{s.data(), static_cast<std::size_t>(slen)};
      const std::uint32_t index = state_->find_path_blockindex(blockindex, comp);
      if (index < 1U) {
        return false;
      }
      blockindex = index;
      if (!state_->add_to_pwdtmp(pathname.size(), comp)) {
        return false;
      }
      slen = 0;
      continue;
    }
    s[static_cast<std::size_t>(slen++)] = pathname[static_cast<std::size_t>(i)];
    if (slen > static_cast<int>(BlockNameMaxSize)) {
      return false;
    }
  }

  if (slen > 0) {
    const std::string_view comp{s.data(), static_cast<std::size_t>(slen)};
    const std::uint32_t index = state_->find_path_blockindex(blockindex, comp);
    if (index < 1U) {
      return false;
    }
    blockindex = index;
    if (!state_->add_to_pwdtmp(pathname.size(), comp)) {
      return false;
    }
  }

  if (state_->tmp.state == 0U) {
    state_->pwd = state_->pwd_tmp;
    state_->pwd_blockindex = blockindex;
  } else {
    state_->tmp.pwd = state_->pwd_tmp;
    state_->tmp.pwd_blockindex = blockindex;
  }
  return true;
}

std::string_view FileSystem::getcwd() const {
  if (!state_) {
    return {};
  }
  if (state_->tmp.state == 0U) {
    return state_->pwd;
  }
  return state_->tmp.pwd;
}

}  // namespace filefs

File: zig/src/pwd.zig
const std = @import("std");
const types = @import("types.zig");
const util = @import("util.zig");
const block_mod = @import("block.zig");

pub fn chdir(self: anytype, pathname: []const u8) bool {
    if (self.fp == null) {
        return false;
    }
    var blockindex: u32 = undefined;
    var start: usize = undefined;
    if (pathname.len > 0 and pathname[0] == '/') {
        blockindex = 1;
        start = 1;
        if (!util.initPwdTmp(self, "/")) return false;
    } else {
        blockindex = util.currentPwdBlockindex(self);
        if (!util.initPwdTmp(self, util.currentPwd(self))) return false;
        start = 0;
    }

    var s: [types.BlockNameMaxSize + 2]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 2);
    var slen: usize = 0;
    var i: usize = start;
    while (i < pathname.len) : (i += 1) {
        if (pathname[i] == '/') {
            if (slen == 0) continue;
            const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
            if (index < 1) return false;
            blockindex = index;
            if (!util.addToPwdTmp(self, s[0..slen])) return false;
            slen = 0;
            continue;
        }
        s[slen] = pathname[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) return false;
    }
    if (slen > 0) {
        const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
        if (index < 1) return false;
        blockindex = index;
        if (!util.addToPwdTmp(self, s[0..slen])) return false;
    }

    if (self.tmp.state == 0) {
        if (!util.bufferSet(self.pwd[0..], &self.pwd_len, self.pwd_tmp[0..self.pwd_tmp_len])) return false;
        self.pwd_blockindex = blockindex;
    } else {
        if (!util.bufferSet(self.tmp.pwd[0..], &self.tmp.pwd_len, self.pwd_tmp[0..self.pwd_tmp_len])) return false;
        self.tmp.pwd_blockindex = blockindex;
    }
    return true;
}

pub fn getcwd(self: anytype) []const u8 {
    if (self.tmp.state == 0) {
        return self.pwd[0..self.pwd_len];
    }
    return self.tmp.pwd[0..self.tmp.pwd_len];
}
